# Contents

#### 01. Importing Libraries and Data
#### 02. Data Consistency Checks for df_prods
    Finding Missing Values
    Addressing Missing Values
    Finding Duplicates
    Addressing Duplicates
#### 03. Data Consistency Checks for df_ords
    Checking for Mixed-Type Data
    Finding Missing Values
    Addressing Missing Values
    Finding Duplicates
#### 04. Exporting Data

# 01. Importing Libraries and Data

In [4]:
import pandas as pd
import numpy as np
import os

In [5]:
# Assign path to str
path = r'/Users/allyfornino/Documents/10-2022 Instacart Basket Analysis/02 Data/'

In [6]:
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'))

In [7]:
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

# 02. Data Consistency Checks for df_prods

## Finding Missing Values

In [8]:
# Counting the number of missing values for each column
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [9]:
# Creating df for all missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [10]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Addressing Missing Values

In [11]:
# Finding original number of rows to compare
df_prods.shape

(49693, 5)

In [12]:
# Creating new df without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
# Checking to see if all missing values were removed
df_prods_clean.shape

(49677, 5)

## Finding Duplicates

In [14]:
# Creating a new df of all duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [15]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## Addressing Duplicates

In [16]:
# Finding original number of rows to compare
df_prods_clean.shape

(49677, 5)

In [17]:
# Creating new df without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [18]:
# Checking to see if all duplicates were removed
df_prods_clean_no_dups.shape

(49672, 5)

In [19]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### Aspects of the data that should be investigated further:
##### Max value of prices = 99999 and mean is 9.99 vs median of 7.10 : 99999 value is possibly used as an indicator of missing values?

# 03. Data Consistency Checks for df_ords

## Checking for Mixed-Type Data

In [20]:
# Checking for mixed-type data in df_ords
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

#### There is no mixed-type data in the df_ords dataframe

## Finding Missing Values

In [21]:
# Counting the number of missing values for each column
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [22]:
# Creating df for all missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [23]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

#### For orders that are the first ones placed by that customer (shown by the order_number value), the 'days_since_prior_order' value is missing because they don't have any prior orders to calculate this value.

## Addressing Missing Values

In [24]:
# Creating a new df with a flag for missing values
df_ords_clean = df_ords

In [25]:
df_ords_clean['new_customer'] = df_ords['order_number'] == 1

In [26]:
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 8 columns]

#### In order to address the missing values, I created a new column called 'new_customer' that acts as a flag for whether or not the customer that placed the order is new, which provides an explanation for the missing 'days_since_prior_order' value.

## Finding Duplicates

In [27]:
# Creating a df of all duplicate rows
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [28]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

#### There are no duplicates in df_ords_clean

In [30]:
df_ords_clean.shape

(3421083, 8)

# 04. Exporting Data

In [ ]:
# Exporting cleaned prods df as products_cleaned.csv
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_cleaned.csv'))

In [ ]:
# Exporting cleaned ords df as orders_cleaned.csv
df_ords_clean.to_csv(os.path.join(path, 'Prepared Data', 'orders_cleaned.csv'))